### ISSUE TITLE
Bug: Vue + TypeScript project slowdown with v8 alpha

### ISSUE THREAD

**Issue by jfrs** [typescript-eslint/typescript-eslint#9312]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

After upgrading a fairly standard Vue + TypeScript project (ESLint from v8 to v9, and `typescript-eslint` from v7 to v8), with a similar configuration the linting time goes from +/- 20 seconds to 3 minutes.

The command is `eslint src`, where  `src` is a directory containing only `.ts` and `.vue` files.

Old `.eslintrc.json`:
```json
{
  "root": true,
  "parser": "vue-eslint-parser",
  "parserOptions": {
    "parser": "@typescript-eslint/parser",
    "project": true,
    "extraFileExtensions": [".vue"]
  },
  "extends": [
    "plugin:@typescript-eslint/strict-type-checked",
    "plugin:@typescript-eslint/stylistic-type-checked",
    "plugin:vue/vue3-recommended",
    "prettier"
  ],
  "overrides": [
    {
      "files": ["src/types/*.d.ts"],
      "rules": {
        "@typescript-eslint/no-extraneous-class": ["error", { "allowConstructorOnly": true }]
      }
    }
  ]
}
```

New `eslint.config.js`
```ts
import eslintConfigPrettier from 'eslint-config-prettier';
import pluginVue from 'eslint-plugin-vue';
import tseslint from 'typescript-eslint';

export default tseslint.config(
  ...tseslint.configs.strictTypeChecked,
  ...tseslint.configs.stylisticTypeChecked,
  ...pluginVue.configs['flat/recommended'],
  eslintConfigPrettier,
  {
    languageOptions: {
      parserOptions: {
        parser: tseslint.parser,
        projectService: true,
        extraFileExtensions: ['.vue']
      }
    }
  },
  {
    rules: {
      '@typescript-eslint/no-unused-vars': ['error', { caughtErrorsIgnorePattern: 'ignore' }]
    }
  },
  {
    files: ['src/types/*.d.ts'],
    rules: {
      '@typescript-eslint/no-extraneous-class': ['error', { allowConstructorOnly: true }]
    }
  },
  {
    ignores: ['eslint.config.js']
  }
);
```

### Reproduction Repository Link

Private repository but I can give people access to it

### Repro Steps

1) clone the repo
2) `pnpm install`
3) `pnpm lint`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `typescript-eslint`     | `8.0.0-alpha.29` |
| `TypeScript`                           | `5.4.5` |
| `ESLint`                               | `9.4.0` |
| `node`                                 | `22.2.0` |

**Comment by higherorderfunctor** [comment]
This may be more to do with ProjectServices than anything.

There is the general [eslint-plugin-import](https://github.com/typescript-eslint/typescript-eslint/issues/9223) issue.  I personally use import-x on my repos and just manually patched it for now.

Also, the way the vue-eslint-parser works, type checking template code destroys performance.  Best [suggestion](https://github.com/vuejs/vue-eslint-parser/issues/104#issuecomment-875149361) I've seen right now is to send to the js parser (espree) and not use this plugin's rules.

I've poked around a little and I think every template `{{ code }}` block is treated as a separate file.  Using tsserver's ProjectService seems to trigger a whole bunch of module resolution and type checking per file.  https://github.com/typescript-eslint/typescript-eslint/issues/8835#issuecomment-2151504601 may also be related to this behavior where files that don't seem to be part of the ProjectService's initialization (which these template blocks may or may not be) are treated as orphaned causing some overhead.

I'm pretty new to tsserver's editor API, but I'm kind of wondering if there is better way to deal with "incremental" project changes in some way to work with vue's behavior and for post initialization changes when using the vscode eslint lsp plugin.

**Comment by higherorderfunctor** [comment]
As was discovered on https://github.com/typescript-eslint/typescript-eslint/issues/8835#issuecomment-2160233861, moving `service.setHostConfiguration` to be set once, probably with the `ProjectService`, seems like low hanging fruit.  Calling it on every file triggers a full project reload and cancels previous synchronizations which cascades with vue template blocks each getting treated as a file.

**Comment by higherorderfunctor** [comment]
I have https://github.com/typescript-eslint/typescript-eslint/pull/9336 ready to go if the maintainers agree with this solution.  Thanks!

**Comment by JoshuaKGoldberg** [comment]
This is fantastic, thanks so much for the deep dives & splitting out @higherorderfunctor!

I'm in-and-out this week and the next for conferences and trying to find the time to get to this ASAP. Apologies if there are delays between reviews - this is high priority IMO.

**Comment by JoshuaKGoldberg** [comment]
Oh and @higherorderfunctor - could you please send me access? My email's on my GitHub profile. I'm happy to sign NDAs / whatever other standard paperwork 😄

**Comment by higherorderfunctor** [comment]
@JoshuaKGoldberg are you looking for access to a test repo or access to my fix branch: https://github.com/higherorderfunctor/typescript-eslint/tree/patches2upstream?

Assuming the former.  We are using enterprise GitLab.  I'll see if I can produce a repo though to play with.

**Comment by higherorderfunctor** [comment]
I setup an example repo that combines [effect](https://github.com/Effect-TS/effect), [vuejs/docs](https://github.com/vuejs/docs), and [AllkindsICP](https://github.com/AllKinds/AllkindsICP) (random repo I found but it uses effect which is heavy on complex types).  I've connected them with project references in the root of the repo.  I tossed in some vue recommend plugin rules to give it something to churn on.  This closely mirrors my work repository with is Effect on the backend with Effect/Vue on the front-end with a very similar monorepo structure using project references.

```js
const vueRules = overrideWith(vueOverrides, [
  ...tsRules,
  ...vueRecommended,
  ...compat.extends("@vue/eslint-config-typescript/recommended"),
  ...compat.extends("plugin:vuetify/recommended"),
  ...vueA11yPlugin.configs["flat/recommended"],
  ...compat.extends("@vue/eslint-config-prettier"),
  { rules: { "codegen/codegen": "off" } } // doesn't work with vue
])
```


The repo is located at (note the branch if cloning): https://github.com/higherorderfunctor/effect/tree/chore/eslint-testing

- I converted the effect's eslint config to be flat.
- It has two patches I added, one for `eslint-plugin-import` via `eslint-module-utils` for not filtering `projectService` and one for the effect's in-house plugin `@effect/eslint-plugin` which did not play nicely with flat configs.
- It is preconfigured for the sum of my patches to `@typescript-eslint/*` via `package.json` which I haven't put all the PRs in for yet: https://github.com/higherorderfunctor/typescript-eslint/tree/patches2upstream.  It will require `pnpm@9` (for git deps) to play with.  You might have to adjust the specific version of `pnpm` at `packageManager` from `9.1.1` to say `9.1.4` or whatever you have installed.

```json
      "@typescript-eslint/types": "higherorderfunctor/typescript-eslint#patches2upstream&path:packages/types",
      "@typescript-eslint/typescript-estree": "higherorderfunctor/typescript-eslint#patches2upstream&path:packages/typescript-estree",
```

To run some tests:

```sh
# re-install deps
pnpm i
```

```sh
# AllkindsICP
/etc/profiles/per-user/caubut/bin/time -v pnpm exec eslint 'packages/AllkindsICP/{tests,src}/**'

✖ 11479 problems (9513 errors, 1966 warnings)
  5626 errors and 1953 warnings potentially fixable with the `--fix` option.

        Elapsed (wall clock) time (h:mm:ss or m:ss): 1:03.49
        Maximum resident set size (kbytes): 2818164
```

```sh
# vue-docs
/etc/profiles/per-user/caubut/bin/time -v pnpm exec eslint 'packages/vue-docs/{tests,src}/**'

✖ 1668 problems (1569 errors, 99 warnings)
  890 errors and 93 warnings potentially fixable with the `--fix` option.

        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:27.11
        Maximum resident set size (kbytes): 1536552
```

```sh
# whole repo
/etc/profiles/per-user/caubut/bin/time -v pnpm lint

✖ 19186 problems (17101 errors, 2085 warnings)
  7551 errors and 2066 warnings potentially fixable with the `--fix` option.

        Elapsed (wall clock) time (h:mm:ss or m:ss): 12:46.26
        Maximum resident set size (kbytes): 4511236
```

To undo my combined `@typescript-eslint/*` patches and just use `rc-v8`, change to these overrides in `package.json`.  Note that I am leaving the `eslint-plugin-import` patch; on my work repo it represented a 4-5x speedup.

```json
      "@typescript-eslint/types": "$typescript-eslint",
      "@typescript-eslint/typescript-estree": "$typescript-eslint",
```

```sh
# install deps
pnpm i
```

```sh
# AllkindsICP
/etc/profiles/per-user/caubut/bin/time -v pnpm exec eslint 'packages/AllkindsICP/{tests,src}/**'
✖ 11479 problems (9513 errors, 1966 warnings)
  5626 errors and 1953 warnings potentially fixable with the `--fix` option.

        Elapsed (wall clock) time (h:mm:ss or m:ss): 36:32.52
        Maximum resident set size (kbytes): 2894116
```

```sh
# vue-docs
/etc/profiles/per-user/caubut/bin/time -v pnpm exec eslint 'packages/vue-docs/{tests,src}/**'

✖ 1668 problems (1569 errors, 99 warnings)
  890 errors and 93 warnings potentially fixable with the `--fix` option.

        Elapsed (wall clock) time (h:mm:ss or m:ss): 15:38.62
        Maximum resident set size (kbytes): 1939532
```

```sh
# whole repo
/etc/profiles/per-user/caubut/bin/time -v pnpm lint

  # not even going to try
```

`vue-docs` had a 97.11% decrease or 34.60x speed up.
`AllkindsICP` had a 97.10% decrease or 34.48x speed up.
The whole repo took 12m46s with the patches.  I'm not even going to try unpatched.  We can guess it will be around ~2h23m.

Below is everything in my combined patch:
- Fixes `extends` not getting processed by the default config used for `projectService` inferred projects. https://github.com/typescript-eslint/typescript-eslint/pull/9306
- Adds tsserver logging. https://github.com/typescript-eslint/typescript-eslint/pull/9337
- Minimizes how many times `setHostConfiguration` is called. https://github.com/typescript-eslint/typescript-eslint/pull/9336
- Adds support to use watches with a Trie lookup (without actual FS watching) to update tsserver with changes.
- Caches open files open files with tsserver and closes them using an LRU based on access order.
- Sends incremental diffs instead of re-opening files if already open.

@JoshuaKGoldberg I planned to break up into several PRs, but due to the time sensitive nature, I can just submit everything in one.  That said, I haven't gotten around to writing tests yet for the ones without a PR.

**Comment by higherorderfunctor** [comment]
This is the current sum of all changes: https://github.com/higherorderfunctor/typescript-eslint/pull/1/files

It is using the branch: https://github.com/higherorderfunctor/typescript-eslint/tree/patches2upstream-no-build-artifacts

This branch is my working branch with the changes but also build artifacts and some pnpm hacks: https://github.com/higherorderfunctor/typescript-eslint/tree/patches2upstream

I'll try and keep my fork's `v8` branch up to do date and make sure changes are applied from `patches2upstream` and `patches2upstream-no-build-artifact`so the diff stays up to date.

**Comment by JoshuaKGoldberg** [comment]
> @JoshuaKGoldberg are you looking for access to a test repo or access to my fix branch

I'm mainly looking for test repos: i.e. I want to see the reproductions. I'd like to play around with as many as possible to understand them. Especially minimum reproductions that don't have a ton of extra stuff.

https://github.com/typescript-eslint/typescript-eslint/issues/9312#issuecomment-2166787001 is an excellent example of this, thanks!

cc @jfrs, anything you can send would be great too!

> planned to break up into several PRs

Heh, yeah, that'd definitely be preferable for us. If you don't have time then we can help with splitting up - but very happy to see that you did send a bunch split out. Thanks! 🙌 

This issue has some of the best reproductions & reporting from multiple people we've gotten this year. Love it!!

**Comment by JoshuaKGoldberg** [comment]
#9336 was merged into `v8`, so marking this issue as closed. But linking to #7906 -> #8031 to note that I'll want to set up performance comparisons for Vue users too. Thanks again all!
